### PR TITLE
Add a taint to the node before draining.

### PR DESCRIPTION
**Pull request by nclaeys**
This makes it possible to detect that the pods are killed due to spot termination for example. Without tainting the node, you have no way of knowing why a job/pod was killed.

We use this information to enrich our error messages shown to the user.

### CODE REVIEW

File: pkg/api/drain.go [hunk 1]
```diff
@@ -16,18 +16,25 @@ import (
...
+	"k8s.io/client-go/util/retry"
 	"k8s.io/kubectl/pkg/drain"
 )
 
 const AzureProviderID = "^azure:///subscriptions/(.+)/resourceGroups/(.+)/providers/Microsoft.Compute/virtualMachineScaleSets/(.+)/virtualMachines/(.+)$" //nolint:lll
 
-var errAzureProviderIDNotValid = errors.New("azureProviderID not valid")
+var (
+	errAzureProviderIDNotValid = errors.New("azureProviderID not valid")
+	taintKeyPrefix             = "aks-node-termination-handler"
```
Comment (maksim-paskal) [inline pkg/api/drain.go:36]: please move `taintKeyPrefix` variable to const

**Comment by maksim-paskal** [comment]
@nclaeys thanks for contributing. I appreciate this. I just want to know what exactly this change can help you. Taint nodes with `TaintEffectNoSchedule` will have no effect on pods that already run on this node. How can this taint help you to identify pod termination reasons?

**Comment by nclaeys** [comment]
Hi Maksim,

I want to thank you providing the node-termination-handler, it looks very good and from initial testing, does a great job. It was one piece we were still missing in Azure vs our AWS offering. 

Well our platform basically monitors batch jobs. We have kubernetes operators to manage them. If a node gets preemted and all the pods get drained, they will receive a sigterm and some of them will fail because of this. We want to show to users, the exact reason why the jobs failed. We do this in our operator by mapping different failure states of our pods/containers to a more human friendly error (translating mounting errors, iam errors, pod pending,...). As part of this we lookup information of the node, and if it has been tainted with preemtion, we know that that is why the job was killed.

It is important to give users this information, since it allows them to decide whether the workloads can run on spot (cheaper) but sometimes they may fail due to preemption. If that is too big an issue, they can run their jobs only on on-demand nodes. 

So actually the most important thing is the node being annotated with a specific key, not the effect of the taint. It helps to distinguish node drains due to upgrading a kubernetes version for example from preemption of spot nodes...

I hope this gives you some more insight, if you have any more questions just shoot.

**Comment by maksim-paskal** [comment]
Thanks for the details. Sounds great. In your case this information can be stored as part of node annotation, but node taint also will be a great option. 

Niels, what about adding new configuration flags `-taint-node=false` and `-taint-effect=NoSchedule` - that will handle adding taints and taint effect before node draining?

**Comment by nclaeys** [comment]
It makes sense to add taint-node parameter and indeed. I could add the taint-effect, but to me this looks less useful, any specific reason why you want people to specify that? Right after the taint you will cordon and drain all the pods anyway so the job will be killed anyway.

**Comment by maksim-paskal** [comment]
It will be great to have customizable `taint-effect` - it will allow for some users terminate also DaemonSets with `NoExecute`  taint effect. For default it will be `NoSchedule`

**Comment by nclaeys** [comment]
Alright fine, I will update the PR to add both. Thanks for the feedback!

File: Makefile [hunk 2]
```diff
@@ -7,8 +7,8 @@ telegramChatID=-439460552
 
 build:
 	git tag -d `git tag -l "helm-chart-*"`
-	go run github.com/goreleaser/goreleaser@latest build --rm-dist --skip-validate --snapshot
-	mv ./dist/aks-node-termination-handler_linux_amd64/aks-node-termination-handler aks-node-termination-handler
+	go1.18.1 run github.com/goreleaser/goreleaser@latest build --rm-dist --skip-validate --snapshot
```
Comment (maksim-paskal) [inline Makefile:10]: please change binnary `go1.18.1` to `go` in this file

File: pkg/api/drain.go [hunk 3]
```diff
@@ -79,13 +94,62 @@ func DrainNode(ctx context.Context, nodeName string) error {
...
+			return false, nodeErr
+		}
+		err = updateNodeWith(ctx, taintKey, taintValue, err, freshNode)
+		switch {
+		case err == nil:
+			return true, nil
+		case apierrorrs.IsConflict(err):
+			return false, nil
+		case err != nil:
+			return false, errors.Wrap(err, fmt.Sprintf("failed to taint node %s with key %s", freshNode.Name, taintKey))
```
Comment (maksim-paskal) [inline pkg/api/drain.go:127]: please use `errors.Wrapf` insted of `errors.Wrap(err, fmt.Sprintf`

File: pkg/api/drain.go [hunk 4]
```diff
@@ -79,13 +94,62 @@ func DrainNode(ctx context.Context, nodeName string) error {
...
+}
+
+func addTaint(ctx context.Context, node *corev1.Node, taintKey string, taintValue string) error {
+	freshNode := node.DeepCopy()
+
+	var err error
+
+	updateErr := wait.ExponentialBackoff(retry.DefaultBackoff, func() (bool, error) {
+		if freshNode, err = Clientset.CoreV1().Nodes().Get(ctx, freshNode.Name, metav1.GetOptions{}); err != nil {
+			nodeErr := errors.Wrap(err, fmt.Sprintf("failed to get node %s", freshNode.Name))
```
Comment (maksim-paskal) [inline pkg/api/drain.go:115]: please use `errors.Wrapf` insted of `errors.Wrap(err, fmt.Sprintf`

File: pkg/api/drain.go [hunk 5]
```diff
@@ -79,13 +94,62 @@ func DrainNode(ctx context.Context, nodeName string) error {
...
+	log.Warnf("Successfully added taint %s on node %s", taintKey, freshNode.Name)
+
+	return nil
+}
+
+func updateNodeWith(ctx context.Context, taintKey string, taintValue string, err error, node *corev1.Node) error {
+	node.Spec.Taints = append(node.Spec.Taints, corev1.Taint{
+		Key:    taintKey,
+		Value:  taintValue,
+		Effect: corev1.TaintEffectNoSchedule,
```
Comment (maksim-paskal) [inline pkg/api/drain.go:146]: change `corev1.TaintEffectNoSchedule` to  `config.Get().TaintEffect`

**Comment by maksim-paskal** [comment]
@nclaeys great job! Thanks for contributing. Your changes released https://github.com/maksim-paskal/aks-node-termination-handler/releases/tag/v1.0.0

**Pull request merged**